lib/compat/wordpress-6.0/class-wp-webfonts.php

```diff
@@ -59,10 +59,78 @@ public function init() {
 		}
 		add_action( $hook, array( $this, 'generate_and_enqueue_styles' ) );
 
+		add_action( 'wp_loaded', array( $this, 'collect_fonts_from_global_styles' ) );
+		add_filter( 'pre_render_block', array( $this, 'collect_fonts_from_block' ), 10, 2 );
+
+		// We are already enqueueing all registered fonts by default when loading the block editor.
+		// So we need to bail out of block and global styles webfont scanning.
+		add_action( 'admin_init', array( $this, 'remove_webfont_scanning_hooks' ) );
+
 		// Enqueue webfonts in the block editor.
 		add_action( 'admin_init', array( $this, 'generate_and_enqueue_editor_styles' ) );
 	}
 
+	public function collect_fonts_from_global_styles() {
+		$global_styles = gutenberg_get_global_styles();
+
+		if ( isset( $global_styles['blocks'] ) ) {
+			// Register used fonts from blocks.
+			foreach ( $global_styles['blocks'] as $setting ) {
+				$font_slug = $this->get_font_slug_from_setting( $setting );
+
+				if ( $font_slug ) {
+					$this->enqueue_font_family_by_slug( $font_slug );
+				}
+			}
+		}
+
+		if ( isset( $global_styles['elements'] ) ) {
+			// Register used fonts from elements.
+			foreach ( $global_styles['elements'] as $setting ) {
+				$this->get_font_slug_from_setting( $setting );
+			}
+		}
+
+		$has_global_typography_setting = isset( $global_styles['typography'] ) && isset( $global_styles['typography']['fontFamily'] );
+
+		if ( ! $has_global_typography_setting ) {
+			return;
+		}
+
+		$font_family_custom = $global_styles['typography']['fontFamily'];
+		$index_to_splice    = strrpos( $font_family_custom, '|' ) + 1;
+		$font_family_slug   = substr( $font_family_custom, $index_to_splice );
+
+		$this->enqueue_font_family_by_slug( $font_family_slug );
+	}
+
+	private function get_font_slug_from_setting( $setting ) {
+		if ( isset( $setting['typography'] ) && isset( $setting['typography']['fontFamily'] ) ) {
+			$font_family = $setting['typography']['fontFamily'];
+
+			// Full string: var(--wp--preset--font-family--slug).
+			// We do not care about the origin of the font, only its slug.
+			preg_match( '/font-family--(?P<slug>.+)\)$/', $font_family, $matches );
+
+			if ( isset( $matches['slug'] ) ) {
+				return $matches['slug'];
+			}
+		}
+	}
```

Comment:
I think these functions might be better placed alongside some of the Global styles code, so that the Webfonts class doesn't need to be aware of what Global styles is, or how it works.

Reply:
As a guiding principle and as the Webfonts API will become a Core API, this API should be agnostic to where the web fonts configurations are coming from. It's job is to register, enqueue, and then generate the font-face CSS for the used web fonts. Agnostic meaning -> it should be usable for themes with and without theme.json and plugins.

Globally styles and theme.json specific settings shouldn't be in the Webfonts API. Instead, I'd suggest moving these specifics to either the theme.json parser or global styles handler. Then modify the web font array of arguments (configuration) before registering it with the API.

Reply:
[See my comments here](https://github.com/WordPress/gutenberg/issues/39332#issuecomment-1077811906). The newer changes in this PR concern me as this PR shifts the API to know too much about global styles and theme.json and moves the API towards only being usable by theme.json themes.

Reply:
Idea: if it doesn't make sense to shift Global styles back to its handler, then another service could be added that sits in between to extract the global styles, orchestrate them into the webfonts configuration array, and then register them with the Webfonts API.

Reply:
Makes total sense. I'll think of alternatives.